library/jwt.c

```diff
@@ -0,0 +1,198 @@
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+};
+
+static int base64url_decode_len(const char *bufcoded) {
+    int nbytesdecoded;
+    register const unsigned char *bufin;
+    register int nprbytes;
+
+    bufin = (const unsigned char *) bufcoded;
+    while (pr2six[*(bufin++)] <= 63);
+
+    nprbytes = (bufin - (const unsigned char *) bufcoded) - 1;
+    nbytesdecoded = ((nprbytes + 3) / 4) * 3;
+
+    return nbytesdecoded;
+}
+
+static int base64url_decode(char *bufplain, const char *bufcoded) {
+    ZITI_LOG(DEBUG, "bufcoded len is: %s", bufcoded);
+
+    register const unsigned char *bufin;
+    register unsigned char *bufout;
+    register int nprbytes;
+
+    bufin = (const unsigned char *) bufcoded;
+    while (pr2six[*(bufin++)] <= 63);
+    nprbytes = (bufin - (const unsigned char *) bufcoded) - 1;
+
+    bufout = (unsigned char *) bufplain;
+    bufin = (const unsigned char *) bufcoded;
+
+    while (nprbytes > 4) {
+        *(bufout++) = (unsigned char) (pr2six[*bufin] << 2 | pr2six[bufin[1]] >> 4);
+        *(bufout++) = (unsigned char) (pr2six[bufin[1]] << 4 | pr2six[bufin[2]] >> 2);
+        *(bufout++) = (unsigned char) (pr2six[bufin[2]] << 6 | pr2six[bufin[3]]);
+        bufin += 4;
+        nprbytes -= 4;
+    }
+
+    if (nprbytes > 1)
+        *(bufout++) = (unsigned char) (pr2six[*bufin] << 2 | pr2six[bufin[1]] >> 4);
+    if (nprbytes > 2)
+        *(bufout++) = (unsigned char) (pr2six[bufin[1]] << 4 | pr2six[bufin[2]] >> 2);
+    if (nprbytes > 3)
+        *(bufout++) = (unsigned char) (pr2six[bufin[2]] << 6 | pr2six[bufin[3]]);
+
+    size_t len = ((char*)bufout - bufplain);
+
+    ZITI_LOG(DEBUG, "base64url_decode len is: %d", len);
+
+    return len;
+}
+
+
+int load_jwt_file(const char *filename, struct enroll_cfg_s *ecfg, ziti_enrollment_jwt_header **zejh, ziti_enrollment_jwt **zej) {
+    ZITI_LOG(DEBUG, "filename is: %s", filename);
+
+    struct stat stats;
+    int s = stat(filename, &stats);
+    if (s == -1) {
+        ZITI_LOG(ERROR, "%s - %s", filename, strerror(errno));
+        return ZITI_JWT_NOT_FOUND;
+    }
+
+    FILE* file = fopen(filename, "r");
+
+    size_t jwt_file_len = (size_t) stats.st_size;
+    ecfg->raw_jwt = calloc(1, jwt_file_len + 1);
+    size_t rc;
+    if ((rc = fread(ecfg->raw_jwt, 1, jwt_file_len, file)) != jwt_file_len) {
+        ZITI_LOG(WARN, "failed to read JWT file in full [%zd/%zd]: %s(%d)", rc, jwt_file_len, strerror(errno), errno);
+    }
+    fclose(file);
+
+    ZITI_LOG(DEBUG, "jwt file content is: \n%s", ecfg->raw_jwt);
+
+    const char *dot1 = strchr(ecfg->raw_jwt, '.');
+    const char *dot2 = strchr(dot1 + 1, '.');
```

Comment:
check for JWT validity: error out if dots were not found

Reply:
added.